lib/Models/sendFeedback.ts

```diff
@@ -0,0 +1,105 @@
+import i18next from "i18next";
+import isDefined from "../Core/isDefined";
+import loadWithXhr from "../Core/loadWithXhr";
+import TerriaError from "../Core/TerriaError";
+import {
+  buildShareLink,
+  buildShortShareLink,
+  canShorten
+} from "../ReactViews/Map/Panels/SharePanel/BuildShareLink";
+import Terria from "./Terria";
+
+export default function sendFeedback(options: {
+  terria: Terria;
+  title?: string;
+  name: string;
+  email: string;
+  sendShareURL: boolean;
+  comment: string;
+  additionalParameters?: Record<string, string | undefined>;
+}) {
+  if (!isDefined(options) || !isDefined(options.terria)) {
+    throw TerriaError.from("options.terria is required.");
+  }
+
+  var terria = options.terria;
+
+  if (!isDefined(terria.configParameters.feedbackUrl)) {
+    raiseError(terria, "`terria.configParameters.feedbackUrl` is not defined");
+    return;
+  }
+
+  const shareLinkPromise = options.sendShareURL
+    ? canShorten(terria)
+      ? buildShortShareLink(terria)
+      : Promise.resolve(buildShareLink(terria))
+    : Promise.resolve("Not shared");
+
+  return shareLinkPromise
+    .then(shareLink => {
+      const feedbackData: Record<string, string | undefined> = {
+        title: options.title,
+        name: options.name,
+        email: options.email,
+        shareLink: shareLink,
+        comment: options.comment
+      };
+      if (
+        options.additionalParameters &&
+        terria.serverConfig.config &&
+        terria.serverConfig.config.additionalFeedbackParameters
+      ) {
+        terria.serverConfig.config.additionalFeedbackParameters.forEach(
+          ({ name }: any) => {
+            feedbackData[name] = options.additionalParameters?.[name];
+          }
+        );
+      }
+      return loadWithXhr({
+        url: terria.configParameters.feedbackUrl,
+        responseType: "json",
+        method: "POST",
+        data: JSON.stringify(feedbackData),
+        headers: {
+          "Content-Type": "application/json"
+        }
+      });
+    })
+    .then(function(json) {
+      if (json instanceof String) {
+        json = JSON.parse(json.toString());
```

Comment:
this is subtly different to the original jsx version, although from a cursory test in the console `JSON.parse(new String('{"a":1}')` seems to work?

Reply:
Yeah I was just trying to please the type checker here. Probably not necessary though.
I'm not sure how we'd even get a `String` here

Reply:
I might leave it, just to be careful